### ISSUE TITLE
Bumping minimum version requirements for Python 3 and GTK 3

### ISSUE THREAD

**Issue by mathiascode** [nicotine-plus/nicotine-plus#1871]
Since Debian 9 is reaching end-of-life in June 2022, and Slackware 15 has recently been released, we have an opportunity to bump our minimum version requirements for Python 3 and GTK 3.

In July 2022, the minimum version requirements for Nicotine+ will change as follows:

- [ ] Python 3.5 -> 3.6
  - Ubuntu 18.04 includes Python 3.6, and we intend to support (popular) LTS distro releases until they reach end of standard support
  - When Ubuntu 18.04 reaches end of standard support in April 2023, we will review if it's feasible to bump the version requirement to 3.7
- [ ] GTK 3.18 -> 3.22
  - The only distro that's still supported and uses GTK 3.18 is Slackware 14.2, while other distros package GTK 3.22. Slackware 14.2 has no EOL date as of now, but we can't realistically support a seven year old GTK version for much longer.

These plans are subject to change.

**Comment by slook** [comment]
It would be worth changing the title of this issue to...

**"Bumping minimum version requirements for Python 3.6 and GTK 3.20"**

... because this is a bit ambiguous, since Nicotine has supported Python 3 for years now.

**Comment by mathiascode** [comment]
> It would be worth changing the title of this issue to...
> 
> **"Bumping minimum version requirements for Python 3.6 and GTK 3.22"**
> 
> ... because this is a bit ambiguous, since Nicotine has supported Python 3 for years now.

My intention was to emphasize version bumps for specific major Python and GTK versions. Sure, Python only has a single supported major version, so it can be dropped from the title, but GTK 3 and GTK 4 are quite different toolkits with separate version requirements.

> I'm curious as to what specific benefit(s) the Gtk version bump to 3.22 will provide, over and above 3.20 ?

As far as I know, no distro uses GTK 3.20 anymore, so it doesn't make sense to provide support for this version. 3.22 is closer to the final state of the GTK 3 series, with 3.24 mainly being a version to introduce some GTK 4 features without larger API breaks.

**Comment by mathiascode** [comment]
Python 3.6 introduced f-strings, which would be a good addition to the codebase when possible. There's a 'consider-using-f-string' rule in Pylint that should help with the migration.

https://www.scivision.dev/python-f-string-speed/

**Comment by mathiascode** [comment]
@slook When are you upgrading from Debian 9 and Python 3.5? I'm thinking about bumping the minimum version requirements in the unstable version fairly soon, since Nicotine+ 3.3.0 likely won't be released until after July (Nicotine+ 3.2.x will still be supported until July).

**Comment by slook** [comment]
> Python 3.6 introduced f-strings

Agreed the f-strings is one of the main highlights, in addition to better performance, it would make constructing status log messages syntactically easier, since the variables are {embedded} so there is no string mapping required which will save on much of the awkward string concatenation :)

> When are you upgrading from Debian 9 and Python 3.5?

Since last month (April 2022) I have been mainly supporting (and therefore running on my own daily driver machines) MX Linux "Wildflower" 21.1+ (based on Debian 11 Bullseye LTS currently with Python 3.9.2-3+) until June 2026, therefore legacy testing is going to require special effort, as it will become a specific task requiring a dedicated system snapshot image and VM...

Options are somewhat limited for fixed release distros that shipped with 3.6.x, I am leaning towards elementary OS 5.1.7 "Hera", or an old Ubuntu distro for this specific purpose.

I feel that the approach of running old legacy installations has served us well in terms of helping us to ensure that Nicotine+ is as bullet proof as possible, I don't know if installing Python 3.6.x onto a recent distro would give us such benefits for testing.

**Comment by mathiascode** [comment]
> Options are somewhat limited for fixed release distros that shipped with 3.6.x, I am leaning towards elementary OS 5.1.7 "Hera", or an old Ubuntu distro for this specific purpose, but I don't know about what GTK versions those are yet.

elementary OS Hera is based on Ubuntu 18.04 LTS, which ships with Python 3.6 and GTK 3.22.30 (which will be the minimum version going forward, based on info from https://repology.org/project/gtk/versions).
